x-pack/packages/observability/observability_utils/es/client/create_observability_es_client.ts

```diff
@@ -25,7 +33,14 @@ export interface ObservabilityElasticsearchClient {
     operationName: string,
     parameters: TSearchRequest
   ): Promise<InferSearchResponseOf<TDocument, TSearchRequest>>;
-  esql(operationName: string, parameters: EsqlQueryRequest): Promise<ESQLSearchResponse>;
+  esql<TDocument = unknown, TSearchRequest extends EsqlQueryRequest = EsqlQueryColumnarRequest>(
+    operationName: string,
+    parameters: TSearchRequest
+  ): Promise<InferESQLResponseOf<TDocument, TSearchRequest>>;
+  esql<TDocument = unknown, TSearchRequest extends EsqlQueryRequest = EsqlQueryRowBasedRequest>(
+    operationName: string,
+    parameters: TSearchRequest
+  ): Promise<InferESQLResponseOf<TDocument, TSearchRequest>>;
```

Comment:
Nope. I struggled with it. With your suggestion, ts can't infer wether it should use  `EsqlQueryColumnarRequest` , if `columnar:true` or `EsqlQueryRowBasedRequest` if `columnar:false`. Perhaps it could be simplified, but I couldn't figure it out.

Reply:
are we using both? I have never used anything other than the default

Reply:
why does this have TDocument?

Reply:
So that the result can be cast to TDoc when not `columnar`

Reply:
What is `TDocument`? Is it the output of the query or the original mapped fields?

Reply:
It's the output of the query.

Reply:
I've made some changes. for parsing the output, the type is now more strict: [535568b](https://github.com/elastic/kibana/pull/198760/commits/535568babf9684eb722d5ed050397e0118402f8a). We need the overloads and generics though